bip-generic-signmessage.mediawiki

```diff
@@ -0,0 +1,134 @@
+<pre>
+  BIP: ????
+  Layer: Applications
+  Title: Generic Signed Message Format
+  Author: Karl-Johan Alm <karljohan-alm@garage.co.jp>
+  Comments-Summary: No comments yet.
+  Comments-URI: https://github.com/bitcoin/bips/wiki/Comments:BIP-????
+  Status: Draft
+  Type: Standards Track
+  Created: 2018-09-10
+  License: CC0-1.0
+</pre>
+
+== Abstract ==
+
+A standard for interoperable generic signed messages based on the Bitcoin Script format.
+
+== Motivation ==
+
+The current message signing standard only works for P2PKH (1...) addresses. By extending it to use a Bitcoin Script based approach, it could be made more generic without causing a too big burden on implementers, who most likely have access to Bitcoin Script interpreters already.
+
+== Specification ==
+
+A new structure <code>SignatureProof</code> is added, which is a simple serializable scriptPubKey, scriptSig & witnessProgram container.
+
+Two actions "Sign" and "Verify" are defined.
+
+=== SignatureProof container ===
+
+{|class="wikitable" style="text-align: center;"
+|-
+!Type
+!Length
+!Name
+!Comment
+|-
+|Uint32||4||flags||standard flags (1-to-1 with standard flags in Bitcoin Core)
+|-
+|VarInt||1-8||msglen||Number of bytes in message string, excluding NUL termination
+|-
+|Char*||[msglen]||msg||The message being signed for all subjects, excluding NUL termination
+|-
+|Uint8||1||entries||Number of proof entries<ref><strong>Why support multiple proofs?</strong> In particular with proof of funds, it is non-trivial to check a large number of individual proofs (one per UTXO) for duplicates. Software could be written to do so, but it seems more efficient to build this check into the specification itself.</ref>
+|}
+
+The above is followed by [entries] number of scriptPubKey & signature entries:
+
+{|class="wikitable" style="text-align: center;"
+|-
+!Type
+!Length
+!Name
+!Comment
+|-
+|VarInt||1-8||spklen||Length of scriptPubKey
+|-
+|Uint8*||[spklen]||spk||The scriptPubKey data
+|-
+|VarInt||1-8||scriptsiglen||Number of bytes in scriptSig data
+|-
+|Uint8*||[scriptsiglen]||scriptsig||ScriptSig data
+|-
+|VarInt||1-8||witlen||Number of bytes in witness program data
+|-
+|Uint8*||[witlen]||wit||Witness program
+|}
+
+In some cases, the scriptsig may be empty (scriptsiglen=0).
+
+=== Signing ===
+
+The "Sign" action takes as input a scriptPubKey and a message (e.g. "hello world"). It succeeds or fails.
+
+# FAIL if scriptPubKey already exists in scriptPubKeys set, otherwise insert it<ref><strong>Why track duplicates?</strong> Because a 3-entry proof is not proving 3 scriptPubKeys unless they are all distinct, or unless they are proving different UTXO:s (see Future Extensions)</ref>
+# Derive the private key privkey for the scriptPubKey, or FAIL
+# Define the message pre-image as the sequence "Bitcoin Message:" concatenated with the message, encoded in UTF-8 using Normalization Form Compatibility Decomposition (NFKD)
+# Let sighash = sha256(sha256(pre-image))
```

Comment:
using plain double-sha256 here is very dangerous, as attackers could trick people to sign real txs this way. You need to either pad something before or after the message, or simply use a different hash function.

for example, it could be `sighash = sha256(sha256(msg)|0xff)`.

If the size of `msg` is small enough to be shown on the wallet screen, the wallet will ask user to confirm the message and will do the whole double hash

If the size of `msg` is too big, the untrusted computer will send sha256(msg) to the wallet, and the wallet will only do the second hash. Careful users will use a trusted machine to verify the hash (they don't need special software because it's simple sha256). Even not verified, the worst case would be signing a wrong message, not sending bitcoin to attacker

Reply:
As for a long message, perhaps there should be a limit on message size. I don't think I understand the benefits of single-sha256'ing and showing that. It doesn't seem like it would convey anything at all to the user.

Reply:
>The msg is prefixed with the string sequence "Bitcoin Message:", so that should not be a problem.

This is 16 bytes at the beginning. To attack it, a 12 bytes (96 bit) collision is needed. Bitcoin mining is now 82bit/day, so it is 45 years for 96 bit. Not likely, but also not impossible.

The only way to make it completely safe (in terms of legacy sighash and BIP143), is to put a non-zero byte at the end of the message.

> I don't think I understand the benefits of single-sha256'ing and showing that. It doesn't seem like it would convey anything at all to the user.
Maybe, but this is what Ledger Wallet is currently doing.

Reply:
@jl2012 

> This is 16 bytes at the beginning. To attack it, a 12 bytes (96 bit) collision is needed. Bitcoin mining is now 82bit/day, so it is 45 years for 96 bit. Not likely, but also not impossible.
> 
> The only way to make it completely safe (in terms of legacy sighash and BIP143), is to put a non-zero byte at the end of the message.

Right now, the sighash is defined as `sha256(sha256(scriptPubKey || pre-image))` where pre-image is "Bitcoin Message:<message string>" for a regular signmessage call. The added scriptPubKey should address the problem, right?

> Maybe, but this is what Ledger Wallet is currently doing.

Interesting. I didn't realize this. I will have to read up on why they do that.